File: 2asir/web-deployment/PHP/clase/OBJETOS/PHP_POO_Mastery_Kit/03_Funciones_Modulos.php
<?php
/**
 * 03_Funciones_Modulos.php
 * 
 * OBJETIVO:
 * Encapsulamiento y miembros estáticos. Proteger los datos y compartir recursos.
 * 
 * CONCEPTOS:
 * 1. Visibilidad:
 *    - public: Accesible desde cualquier lugar.
 *    - private: Solo accesible DESDE LA PROPIA CLASE.
 *    - protected: Solo clase y herencia (lo veremos luego).
 * 2. Getters y Setters: Métodos públicos para controlar propiedades privadas.
 * 3. Static: Propiedades/Métodos que pertenecen a la CLASE, no al objeto.
 */

class Usuario
{
    // PRIVATE: Nadie puede tocar esto directamente desde fuera
    private $password;
    public $username;

    // STATIC: Contador compartido por TODOS los usuarios
    public static $totalUsuarios = 0;

    public function __construct($nombre, $pass)
    {
        $this->username = $nombre;
        $this->setPassword($pass); // Usamos el setter interno

        // Aumentamos el contador de la CLASE (self::, no $this->)
        self::$totalUsuarios++;
    }

    // SETTER: Validamos antes de guardar
    public function setPassword($passNueva)
    {
        if (strlen($passNueva) < 6) {
            echo "[ERROR] La contraseña para {$this->username} es muy corta.\n";
        } else {
            // Simulamos encriptación
            $this->password = hash('sha256', $passNueva);
            echo "[OK] Contraseña actualizada correctamente.\n";
        }
    }

    // GETTER: Devuelve un dato (pero nunca la contraseña real)
    public function getPasswordHash()
    {
        return $this->password;
    }

    // MÉTODO ESTÁTICO: Se puede llamar sin crear objetos
    public static function verContador()
    {
        return "Actualmente hay " . self::$totalUsuarios . " usuarios registrados.";
    }
}

## -----------------------------------------
## SECCIÓN DE PRUEBAS DE ENCAPSULAMIENTO
## -----------------------------------------

echo "Usuarios iniciales: " . Usuario::verContador() . "\n\n";

$user1 = new Usuario("Ana", "123"); // Error, corta
$user1->setPassword("secreto123");   // OK

// Intentar acceder directamente a privado da error fatal:
// echo $user1->password; // ESTO FALLARÍA

echo "Hash guardado: " . $user1->getPasswordHash() . "\n";

## -----------------------------------------
## SECCIÓN DE PRUEBAS ESTÁTICAS
## -----------------------------------------

$user2 = new Usuario("Carlos", "passwordSegura");
$user3 = new Usuario("Eva", "claveMaestra");

// El contador ha subido solo
echo "\n" . Usuario::verContador() . "\n";

/*
 * NOTA DOCENTE:
 * La encapsulación (private/getters/setters) es el pilar de la seguridad en POO.
 * "Static" es útil para contadores globales o utilidades que no dependen de un objeto específico.
 */
?>

File: 2asir/web-deployment/PHP/clase/OBJETOS/Python_POO_Mastery_Kit/03_Funciones_Modulos.py
"""
03_Funciones_Modulos.py

OBJETIVO:
Encapsulamiento y estáticos.

DIFERENCIAS ENORME CON PHP:
1. Python NO tiene 'private' o 'protected' reales. Todo es público por defecto.
2. Convención: Si una variable empieza por `_` (ej. `_password`), SE RESPETA como privada, aunque técnicamente podrías tocarla.
3. Estáticos: Se usan decoradores `@staticmethod` o `@classmethod`.
"""

import hashlib # Necesario para simular el hash

class Usuario:
    # Atributo DE CLASE (Estático en concepto)
    total_usuarios = 0
    
    def __init__(self, nombre, password):
        self.username = nombre
        # El guion bajo `_` indica "NO TOQUES ESTO DESDE FUERA"
        self._password = "" 
        
        self.set_password(password)
        
        # Aumentamos el contador de la CLASE
        Usuario.total_usuarios += 1
        
    def set_password(self, pass_nueva):
        if len(pass_nueva) < 6:
            print(f"[ERROR] La contraseña para {self.username} es muy corta.")
        else:
            # En Python usamos hashlib para sha256
            self._password = hashlib.sha256(pass_nueva.encode()).hexdigest()
            print("[OK] Contraseña actualizada correctamente.")
            
    def get_password_hash(self):
        return self._password
    
    # MÉTODO ESTÁTICO (Decorador @staticmethod)
    # No recibe 'self' porque no depende de un objeto concreto
    @staticmethod
    def ver_contador():
        return f"Actualmente hay {Usuario.total_usuarios} usuarios registrados."

## -----------------------------------------
## PRUEBAS
## -----------------------------------------

print(f"Usuarios iniciales: {Usuario.ver_contador()}\n")

user1 = Usuario("Ana", "123")        # Error
user1.set_password("secreto123")     # OK

# En Python PODRÍAS hacer print(user1._password) y funcionaría,
# pero un buen programador Python NUNCA lo haría.
print(f"Hash guardado: {user1.get_password_hash()}")

## ESTÁTICOS
user2 = Usuario("Carlos", "passwordSegura")
user3 = Usuario("Eva", "claveMaestra")

print(f"\n{Usuario.ver_contador()}")
